src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey_slh_dsa.c

```diff
@@ -0,0 +1,280 @@
+    {
+        if (dstLen != Int32ToSizeT(destinationLen))
+        {
+            assert(false); // length mismatch
+            goto done;
+        }
+
+        ret = 1;
+    }
+    else
+    {
+        ret = 0;
+    }
+
+done:
+    if (ctx != NULL) EVP_PKEY_CTX_free(ctx);
+    return ret;
+#else
+    (void)pkey;
+    (void)extraHandle;
+    (void)msg;
+    (void)msgLen;
+    (void)context;
+    (void)contextLen;
+    (void)destination;
+    (void)destinationLen;
+    return -1;
+#endif
+}
+
+int32_t CryptoNative_SlhDsaVerifyPure(EVP_PKEY *pkey,
+                                      void* extraHandle,
+                                      uint8_t* msg, int32_t msgLen,
+                                      uint8_t* context, int32_t contextLen,
+                                      uint8_t* sig, int32_t sigLen)
+{
+    assert(pkey);
+    assert(msgLen >= 0);
+    assert(sig);
+    assert(sigLen >= 7856 /* SLH-DSA-SHA2-128s/SLH-DSA-SHAKE-128s signature size */);
+    assert(contextLen >= 0);
+
+#if defined(NEED_OPENSSL_3_0) && HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return -1;
+    }
+
+    ERR_clear_error();
+
+    EVP_PKEY_CTX* ctx = NULL;
+
+    int ret = -1;
+
+    ctx = EvpPKeyCtxCreateFromPKey(pkey, extraHandle);
+    if (!ctx)
+    {
+        goto done;
+    }
+
+    OSSL_PARAM contextParams[] =
+    {
+        OSSL_PARAM_construct_end(),
+        OSSL_PARAM_construct_end(),
+    };
+
+    if (context)
+    {
+        contextParams[0] = OSSL_PARAM_construct_octet_string(OSSL_SIGNATURE_PARAM_CONTEXT_STRING, (void*)context, Int32ToSizeT(contextLen));
+    }
+
+    if (EVP_PKEY_verify_message_init(ctx, NULL, contextParams) <= 0)
+    {
+        goto done;
+    }
+
+    ret = EVP_PKEY_verify(ctx, sig, Int32ToSizeT(sigLen), msg, Int32ToSizeT(msgLen)) == 1;
+
+done:
+    if (ctx != NULL) EVP_PKEY_CTX_free(ctx);
+    return ret;
+#else
+    (void)pkey;
+    (void)extraHandle;
+    (void)msg;
+    (void)msgLen;
+    (void)context;
+    (void)contextLen;
+    (void)sig;
+    (void)sigLen;
+    return -1;
+#endif
+}
+
+int32_t CryptoNative_SlhDsaExportSecretKey(const EVP_PKEY* pKey, uint8_t* destination, int32_t destinationLength)
+{
+    return EvpPKeyGetKeyOctetStringParam(pKey, OSSL_PKEY_PARAM_PRIV_KEY, destination, destinationLength);
+}
+
+int32_t CryptoNative_SlhDsaExportPublicKey(const EVP_PKEY* pKey, uint8_t* destination, int32_t destinationLength)
+{
+    return EvpPKeyGetKeyOctetStringParam(pKey, OSSL_PKEY_PARAM_PUB_KEY, destination, destinationLength);
+}
+
+int32_t CryptoNative_SlhDsaGetPalId(const EVP_PKEY* pKey, int32_t* slhDsaTypeId)
+{
+#ifdef NEED_OPENSSL_3_0
+    assert(pKey && slhDsaTypeId);
+
+    if (API_EXISTS(EVP_PKEY_is_a))
+    {
+        ERR_clear_error();
+
+        // TODO: This conditional chain seems unavoidable. If there are multiple synonyms for a given key,
```

Comment:
I don't think this needs to be TODO, this is only called when user uses external key handle which is pretty rare, most likely the key is private key which is not supposed to be read more than once and we're only comparing couple of strings here so I doubt this will ever have any perf impact on anything